pkg/event/queue.go

```diff
@@ -19,60 +19,85 @@ package event
 
 import (
 	"sync"
+
+	"github.com/optimizely/go-sdk/pkg/logging"
 )
 
 // Queue represents a queue
 type Queue interface {
-	Add(item interface{})
+	Add(item interface{}) // Should return a bool
 	Remove(count int) []interface{}
 	Get(count int) []interface{}
 	Size() int
 }
 
 // InMemoryQueue represents a in-memory queue
 type InMemoryQueue struct {
-	Queue []interface{}
-	Mux   sync.Mutex
+	logger  logging.OptimizelyLogProducer
+	MaxSize int
+	Queue   []interface{}
+	Mux     sync.Mutex
 }
 
 // Get returns queue for given count size
-func (i *InMemoryQueue) Get(count int) []interface{} {
-	if i.Size() < count {
-		count = i.Size()
-	}
-	i.Mux.Lock()
-	defer i.Mux.Unlock()
-	return i.Queue[:count]
+func (q *InMemoryQueue) Get(count int) []interface{} {
+	q.Mux.Lock()
+	defer q.Mux.Unlock()
+
+	count = q.getSafeCount(count)
+	return q.Queue[:count]
 }
 
 // Add appends item to queue
-func (i *InMemoryQueue) Add(item interface{}) {
-	i.Mux.Lock()
-	i.Queue = append(i.Queue, item)
-	i.Mux.Unlock()
+func (q *InMemoryQueue) Add(item interface{}) {
+	q.Mux.Lock()
+	defer q.Mux.Unlock()
+
+	if len(q.Queue) >= q.MaxSize {
+		q.logger.Warning("MaxQueueSize has been met. Discarding event")
+		return
```

Comment:
Shouldn't this be returning error or something? Or is that planned in a follow up?

Reply:
We should yes, and this is what prompted me to add that comment on the struct to return a boolean. So for now planned follow-up.